src/Pool.ts

```diff
@@ -525,14 +525,15 @@ export class Pool<RawResource> {
       using === this._inUseObjects.length
     ) {
       this._ensureMinimum();
-      return;
+      return Promise.resolve();
     }
 
     this._count -= 1;
     if (this._count < 0) this._count = 0;
 
-    this._factory.destroy(resource);
-    this._ensureMinimum();
+    return Promise.resolve(this._factory.destroy(resource)).then(() => {
```

Comment:
Since `resource` was removed from `availableObjects`, I suggest use of `finally` to ensure minimum number of resources even in case of failure